planet/src/main/java/com/google/planet/data/ItineraryGenerator.java

```diff
@@ -123,14 +117,56 @@ private int getTravelDurationInMinutes(Duration duration) {
                 items.add(item);
 
                 // Update the next event's start time
-                if (i != events.size() - 1) { 
-                    start += event.getDurationInMinutes() + getTravelDurationInMinutes(travelTimeGraph[i][i+1]);
-                }
+                start += event.getDurationInMinutes() + getTravelDurationInMinutes(directionsLegs[i].duration);
             }else {
                 errorMessage = "Sorry, you have too many events in a day!";
                 return Arrays.asList();
             }
         }
         return items;
     }
+
+    // Function that makes a request to the Directions API and return a direction route
+    // A DirectionsRoute has field legs, which is an array of DirectionsLegs
+    // Each DirectionsLeg contains starting address, ending address, and duration
+    // DirectionRoute also has int[] waypointOrder, which will be empty if optimized is set to false,
+    // and will contain the new order of waypoints if optimized is set to true
+    // For the MVP, real time traffic is NOT used.
+    private DirectionsRoute getDirectionsRoute(List<Event> events, boolean optimized) {
+        String GoogleApiKey = "AIzaSyDK36gDoYgOj4AlbCqh1IuaUuTlcpKF0ns";
+        String origin = events.get(0).getAddress();
+        String destination = origin; // The ending location should be assumed as the starting location 
+                                     // since users most likely want a round trip
+        String[] waypoints = getListOfWaypoints(events);
+        GeoApiContext context = new GeoApiContext.Builder(new GaeRequestHandler.Builder())
+                            .apiKey(GoogleApiKey)
+                            .build();
+        DirectionsApiRequest directionsRequest = DirectionsApi.newRequest(context);
+       try {
+            DirectionsResult directionsResult =  directionsRequest
+                                                .origin(origin)
+                                                .destination(destination)
+                                                .waypoints(waypoints)
+                                                .optimizeWaypoints(optimized)
+                                                .await();
+            DirectionsRoute directionsRoute = directionsResult.routes[0];
+            return directionsRoute;
+        } catch (Exception e) {
+            errorMessage = "Oops, one of your addresses is invalid, please use a valid address.";
```

Comment:
Is there a way to know which address is invalid?

Reply:
unfortunately no :( 
When one of the addresses is invalid the api just returns an empty route

Reply:
Okay